Support noise scaling by layer

vprusso opened the pull request:
Closes #1722 

<!--
If the validation checks fail
  1. Run `make check-types` (from the root directory of the repository) and fix any mypy (https://mypy.readthedocs.io/en/stable/) errors.

  2. Run `make check-style` and fix any flake8 (http://flake8.pycqa.org) errors.

  3. Run `make format` to format your code with the black (https://black.readthedocs.io/en/stable/index.html) autoformatter.

For more information, check the Mitiq style guidelines (https://mitiq.readthedocs.io/en/stable/contributing.html#style-guidelines).
-->

## Description

1. Adding a feature to fold circuit by layer. The function `layer_folding` allows one to provide a `cirq.Circuit` object and supply a list of integers of size equal to the depth of the circuit where each integer in the list corresponds to how many folds to apply on a given layer. 

2. Adding corresponding unit tests for folding functionality.

3. Adding documentation for layerwise folding.
---

### License

- [x] I license this contribution under the terms of the GNU GPL, version 3 and grant Unitary Fund the right to provide additional permissions as described in section 7 of the GNU GPL, version 3.

Before opening the PR, please ensure you have completed the following where appropriate.
- [x] I added unit tests for new code.
- [x] I used [type hints](https://www.python.org/dev/peps/pep-0484/) in function signatures.
- [x] I used [Google-style](https://google.github.io/styleguide/pyguide.html#383-functions-and-methods) docstrings for functions.
- [x] I [updated the documentation](../blob/master/docs/CONTRIBUTING_DOCS.md) where relevant.

natestemen commented on mitiq/zne/scaling/layer_scaling.py:
```diff
@@ -0,0 +1,74 @@
+
+"""Functions for layer-wise unitary folding on supported circuits."""
+from typing import List
+from cirq import Circuit, inverse, Moment
```
To reduce typing ambiguity what do you think about something like this?
```suggestion
import cirq
from cirq import inverse, Moment
```
This way in type hints we will use `cirq.Circuit`, as other frontends can call their abstraction `Circuit` as well. Pretty minor, however. WDYT?

natestemen commented on mitiq/zne/scaling/layer_scaling.py:
```diff
@@ -0,0 +1,74 @@
+from cirq import Circuit, inverse, Moment
+
+
+def layer_folding(circuit: Circuit, layers_to_fold: List[int]) -> Circuit:
```
Do you think we should apply [`noise_scaling_converter`](https://github.com/unitaryfund/mitiq/blob/c849314bc73c2e7b1b71d312639c5e8417ab2e74/mitiq/interface/conversions.py#L221-L223) here as a decorator? It should allow for this function to be used with other frontends.

If we decide to add this, we should test with an additional frontend as well.

natestemen replied:
For consistency with the [folding](https://github.com/unitaryfund/mitiq/blob/c849314bc73c2e7b1b71d312639c5e8417ab2e74/mitiq/zne/scaling/folding.py#L219-L223) methods, and the [identity insertion](https://github.com/unitaryfund/mitiq/blob/c849314bc73c2e7b1b71d312639c5e8417ab2e74/mitiq/zne/scaling/identity_insertion.py#L82) scaling method, what if we separated `layers_to_fold` into two separate variables:

1. `layers_to_fold`: To record layer indices
2. `scale_factors`: to record the number of times to fold

I do like how the current API is very dense in it's input data. Not a clear winner for me here, but I lean slightly towards having consistent API if it doesn't seem like a disadvantage to you.

natestemen commented on mitiq/zne/scaling/layer_scaling.py:
```diff
@@ -0,0 +1,74 @@
+    """
+    layers = []
+    for i, layer in enumerate(circuit):
+        layers.append(layer)
+        # Apply the requisite number of folds to each layer.
+        num_fold = layers_to_fold[i]
+        for _ in range(num_fold):
+            layers.append(Moment(inverse(layer)))
+            layers.append(Moment(layer))
```
Is there a reason we need to call `Moment` in the two append calls, but not at the top? My local testing indicates they might be unnecessary.

vprusso replied:
`mypy` seems to throw a fit if this isn't explicitly stated here. Removing the `Moment()` wrapper and running `make check-types` yields:

```
Argument 1 to "append" of "list" has incompatible type "Union[Operation, OpTree]"; expected "Moment"  [arg-type]
````

We could also explicitly ignore the line as well. Any preference?

natestemen replied:
Weird, is it complaining about appending the layer, or the inverse (or both)? From my testing they should all be moments!

```py
c = cirq.Circuit()
q = cirq.LineQubit.range(2)

c.append(cirq.ops.H.on_each(*q))
c.append(cirq.ops.X.on_each(*q))

for layer in c:
    print(type(layer))
    print(type(cirq.inverse(layer)))

>>> <class 'cirq.circuits.moment.Moment'>
>>> <class 'cirq.circuits.moment.Moment'>
>>> <class 'cirq.circuits.moment.Moment'>
>>> <class 'cirq.circuits.moment.Moment'>
```

What does [`reveal_type(layer)`](https://mypy.readthedocs.io/en/stable/type_inference_and_annotations.html?highlight=reveal_type#type-inference) show?

vprusso replied:
Yeah, it's complaining about both the layer and the inverse. Indeed, when I check the types of the layer and inverse, I see `Moment` as well. I think this might be a quirk of `mypy` or perhaps the version that is pinned is getting confused as without this explicit typecast it's failing in the pipeline.

natestemen replied:
WDYT of this?

```py
if not cirq.is_measurement(layer):
    layers.append(layer**-1)
    layers.append(layer)
```

The advantage of using [`cirq.inverse`](https://github.com/quantumlib/Cirq/blob/v1.1.0/cirq-core/cirq/protocols/inverse_protocol.py#L70-L121) seems to be that it ensures `__pow__` is defined, but I believe this should always be the case for moments that are not measurements, right?

natestemen commented on mitiq/zne/scaling/layer_scaling.py:
```diff
@@ -0,0 +1,74 @@
+    # We combine each layer into a single circuit.
+    combined_circuit = Circuit()
+    for layer in layers:
+        combined_circuit.append(Circuit(layer))
```
Similar to above, is this `Circuit(layer)` call needed, or can we append `layer` directly?

vprusso replied:
This seems to be something we can do. Change made.

natestemen commented on mitiq/zne/scaling/layer_scaling.py:
```diff
@@ -0,0 +1,74 @@
+    return combined_circuit
+
+
+def layer_folding_all(
+    circuit: Circuit, layers_to_fold: int = 1
+) -> List[Circuit]:
```
```suggestion
def layer_folding_all(
    circuit: Circuit, num_folds: int = 1
) -> List[Circuit]:
```

Might be a better name?

vprusso replied:
Agreed. Made this change. Thank you!

natestemen commented on docs/source/guide/zne-3-options.md:
```diff
@@ -216,6 +216,66 @@ folding method. Fidelity values must be between zero and one.
+$L$ contains one or more quantum gates acting concurrently on an $n$-qubit
+system:
+
+$$C = L_d L_{d-1} \ldots L_2 L_1$$.
```
Potential stylistic preference, but `\cdots` seems more appropriate here since it's multiplication.

vprusso replied:
Done!

natestemen commented on docs/source/guide/zne-3-options.md:
````diff
@@ -216,6 +216,66 @@ folding method. Fidelity values must be between zero and one.
+For example, consider the following circuit
+
+```{code-cell} ipython3
+from mitiq.zne.scaling.layer_scaling import layer_folding
````
```suggestion
from mitiq.zne.scaling import layer_folding
```

I think we should be able to import directly from `scaling` now.

vprusso replied:
I would think so as well, but attempting to import this seems to yield an error when running on the pipeline for mitiq. I'm keeping the longer import style here, but let me know if you think it's possible to use the shorter one without the pipeline failure.

natestemen replied:
Is there a build where this was tested and it failed? I see https://github.com/unitaryfund/mitiq/pull/1767/commits/a547bc285e14c7b488a2d57b5c7187736d1b1ba2, but when inspecting the build, it seems all the failures I saw was from an unrelated error that was fixed elsewhere.

vprusso replied:
*Think* this should be okay now.

natestemen reviewed (commented):
Really nice work Vincent! Thanks for studying this technique, and getting it worked in so quickly.

I've left a few comments, most of which are completely optional, and I of course will not be offended if you don't agree with a suggestion. My bigger question is do we want this PR to add capability to use this layer folding technique within ZNE? I _think_ it's very close, and so I lean yes, but I have not tested it yet, and there may be unforeseen limitations/challenges. Let me know what you think!

natestemen commented:
Thanks for the nice chat today Vincent. As discussed, I think this PR should aim to implement a folding method that can be used within ZNE. What does this look like in practice (modulo function naming)?

```py
layer_to_fold = 999
fold_layer_func = mitiq.zne.scaling.fold_layer(layer_to_fold)

mitiq.zne.execute_with_zne(circuit, executor, scale_noise=fold_layer_func)
```

The `scale_noise` parameter has the following signature: https://github.com/unitaryfund/mitiq/blob/d6dc058384b5b2b8eac76d448ae13d3dbf0d8e6f/mitiq/zne/zne.py#L31-L33

Hence, however we go about this (refactoring one of the existing functions, or creating a new one), we need a function with the following signature:
```
Callable[
    [int], # layer
    Callable[[
            QPROGRAM, # circuit to fold
            float # scale factor
        ],
        QPROGRAM # folded circuit
    ]
]
# in one line for readability:
Callable[[int], Callable[[QPROGRAM, float], QPROGRAM]]
```

Does this make sense? Let me know if you have any questions, or disagree that this should be the approach.

nathanshammah commented on docs/source/guide/zne-3-options.md:
```diff
@@ -216,6 +216,66 @@ folding method. Fidelity values must be between zero and one.
+This circuit consists of two layers (the first layer has a Hadamard gate and the
+second layer has a CNOT gate).
+
+We can make use of the `layer_folding` function where the argument
```
```suggestion
We can make use of the {func}`.layer_folding()` function where the argument
```

nathanshammah commented on docs/source/guide/zne-3-options.md:
```diff
@@ -216,6 +216,66 @@ folding method. Fidelity values must be between zero and one.
+$L$ contains one or more quantum gates acting concurrently on an $n$-qubit
+system:
+
+$$C = L_d L_{d-1} \cdots L_2 L_1$$.
+
+The $i$-inverse operation of a circuit $C$ is defined as
```
Something is off here:
<img width="516" alt="latex-bug-mitiq" src="https://user-images.githubusercontent.com/14573436/230585635-c9656705-328f-4a35-ba8e-76c493ed3d7a.png">
Note you can preview the docs relative to this PR by clicking on the CI check for readthedocs. Direct link [here](https://mitiq--1767.org.readthedocs.build/en/1767/).

vprusso replied:
Hmm. Oddly enough, this is rendering fine for me locally and I'm not seeing any obvious LaTeX issues with the source. Maybe on the next build it will fix itself. Will see.

vprusso replied:
Resolved. Thanks again @natestemen

nathanshammah reviewed (changes requested):
@vprusso looks very neat to me, really excited to have this in.
I added some small comments on the documentation.
- Please also add the new module to the API doc file, and include a subtitle like for the identity insertion scaling, [link](https://mitiq--1767.org.readthedocs.build/en/1767/apidoc.html#module-mitiq.zne.scaling.identity_insertion).

vprusso commented:
> @vprusso looks very neat to me, really excited to have this in. I added some small comments on the documentation.
> 
> * Please also add the new module to the API doc file (`apidoc.md`), and include a subtitle like for the identity insertion scaling, [link](https://mitiq--1767.org.readthedocs.build/en/1767/apidoc.html#module-mitiq.zne.scaling.identity_insertion).

Added to `apidoc.md`. Nice catch!

natestemen commented on docs/source/guide/zne-3-options.md:
```diff
@@ -216,6 +216,72 @@ folding method. Fidelity values must be between zero and one.
+contains one or more quantum gates acting concurrently on an $n$-qubit
+system:
+
+$$
+C = L_d L_{d-1} \cdots L_2 L_1
+$$.
```
Looks like the error is caused by the period _after_ the `$$`. If you move it directly inside it works.

```suggestion
$$C = L_d L_{d-1} \cdots L_2 L_1.$$
```

vprusso replied:
Excellent.  Saved me some headache there. Thank you! :)

andreamari commented:
@vprusso, for some reason the diff of this PR contains a lot of unrelated previous PRs. Could you try merging master on this PR, or rebasing this PR on the latest Mitiq master?

@natestemen: I am not sure there is a simple way of casting `layer_folding` as a standard noise scaling function compatible with `execute_with_zne`. They have a different number of input parameters. Even if we decide to quantify the amount of noise scaling with a `scale_factor` instead of `num_folds`, the main problem is that `layer_folding` wants a lot of scale factors (one for each layer).

Maybe we should consider `layer_folding` more like a self-contained tool that the user can apply for circuit debugging or for a low-level custom implementation of ZNE.

Note also that what is now called `layer_folding_all` has the same effect as local folding (`fold_gates_at_random`) which, with odd integer scale factors, scales all the gates (and therefore all the layers) in the same way of `layer_folding_all`. So I am not sure if it's useful in the current form.

vprusso commented:
> @vprusso, for some reason the diff of this PR contains a lot of unrelated previous PRs. Could you try merging master on this PR, or rebasing this PR on the latest Mitiq master?

Hmm. That's odd. I merged the `master` branch into the PR branch and it looks okay with the exception of one doc failure. Not entirely sure why this is failing (maybe it's obvious and either @natestemen or @andreamari has seen this one before?)
https://github.com/unitaryfund/mitiq/actions/runs/4692483610/jobs/8318255429#step:6:881

andreamari commented:
> Hmm. That's odd. I merged the `master` branch into the PR branch and it looks okay with the exception of one doc failure. Not entirely sure why this is failing (maybe it's obvious and either @natestemen or @andreamari has seen this one before?) https://github.com/unitaryfund/mitiq/actions/runs/4692483610/jobs/8318255429#step:6:881

We can ignore the error. (We check for broken links in CI and a link to a paper was temporarily broken)

natestemen commented:
I think it is a straightforward change that makes this change more useful, faster. I implemented (quickly) what I had in mind in my [previous comment](https://github.com/unitaryfund/mitiq/pull/1767#issuecomment-1498177636) here: https://github.com/unitaryfund/mitiq/commit/c80124a6369d400949c0609152d6b2efe6742655. Feel free to cherry pick it, or modify as you see fit. It's a relatively small diff, but definitely needs some more testing. It should work as I outlined above, and hopefully with `execute_with_zne` as well, but I did not test that yet.

andreamari commented:
> I think it is a straightforward change that makes this change more useful, faster. I implemented (quickly) what I had in mind in my [previous comment](https://github.com/unitaryfund/mitiq/pull/1767#issuecomment-1498177636) here: 
[c80124a](https://github.com/unitaryfund/mitiq/commit/c80124a6369d400949c0609152d6b2efe6742655). Feel free to cherry pick it, or modify as you see fit. It's a relatively small diff, but definitely needs some more testing. It should work as I outlined above, and hopefully with `execute_with_zne` as well, but I did not test that yet.

If we implement something like [c80124a](https://github.com/unitaryfund/mitiq/commit/c80124a6369d400949c0609152d6b2efe6742655) (I like it), can we call `fold_layer` as `get_layer_folding` or `fold_layer_function`, or anything which gives the idea of returning a folding function instead of actually applying the folding operation?

@natestemen and @vprusso, based on this discussion, do you agree that `layer_folding_all()` can be removed since it is  equivalent to `fold_gates_at_random` up to a `num_folds` --> odd `scale_factor` conversion?

vprusso commented:
> > I think it is a straightforward change that makes this change more useful, faster. I implemented (quickly) what I had in mind in my [previous comment](https://github.com/unitaryfund/mitiq/pull/1767#issuecomment-1498177636) here:
> > [c80124a](https://github.com/unitaryfund/mitiq/commit/c80124a6369d400949c0609152d6b2efe6742655). Feel free to cherry pick it, or modify as you see fit. It's a relatively small diff, but definitely needs some more testing. It should work as I outlined above, and hopefully with `execute_with_zne` as well, but I did not test that yet.
> 
> If we implement something like [c80124a](https://github.com/unitaryfund/mitiq/commit/c80124a6369d400949c0609152d6b2efe6742655) (I like it), can we call `fold_layer` as `get_layer_folding` or `fold_layer_function`, or anything which gives the idea of returning a folding function instead of actually applying the folding operation?
> 
> @natestemen and @vprusso, based on this discussion, do you agree that `layer_folding_all()` can be removed since it is equivalent to `fold_gates_at_random` up to a `num_folds` --> odd `scale_factor` conversion?

- Incorporating/cherry-picking @natestemen  suggestion with the name change as suggested by @andreamari 
- Removing `fold_layer_all` function as suggested by @andreamari 
- Tests for layerwise folding decorator in `test_zne.py`

Let me know what you both think, and specifically, if the tests at present make sense.

vprusso commented:
An additional question is regarding the typing of things. the `get_layer_folding` function takes in a `scale_factor` argument that is a float, but the way in which this argument is used in `layer_folding` only makes sense if the argument is an integer. I wonder if we want to either rethink the way that we perform the folding or treat the scale factors for this instance. Any thoughts, @natestemen / @andreamari ?

natestemen commented on mitiq/zne/scaling/layer_scaling.py:
```diff
@@ -0,0 +1,76 @@
+
+def get_layer_folding(layer_index: int) -> Callable[[QPROGRAM, float], QPROGRAM]:
+    @noise_scaling_converter
+    def fold_ith_layer(
+        circuit: cirq.Circuit, scale_factor: float
+    ) -> cirq.Circuit:
```
```suggestion
    def fold_ith_layer(
        circuit: cirq.Circuit, scale_factor: int
    ) -> cirq.Circuit:
```

Since `layer_folding` doesn't currently support float scale factors, lets make this an int for now.

vprusso replied:
Makes sense. Removed and adapted as suggested!

andreamari commented on mitiq/zne/scaling/layer_scaling.py:
```diff
@@ -0,0 +1,78 @@
+        circuit: The input cirq circuit.
+        layers_to_fold: A list with the index referring to the layer number,
+                        and the element filled by an integer represents the
+                        number of times the layer is folded.
+
```
There is a small ambiguity about the length of `layers_to_fold`: do we count measurements as a layer?
Based on the code the answer seems to be :  yes, but the associated value should be 0 since in practice we cannot fold measurement layers.

Another option is to copy the workflow used in other folding functions where `check_foldable` and `pop_measurements` and `append_measurements` are called. In this case, `layers_to_fold` should be defined without considering the measurement layer (if present).

Whatever approach we use, we should probably say in the docstring what is the expected structure of `layers_to_fold` when measurements are present in the circuit.

vprusso replied:
I think I like the second option you proposed here for the sake of consistency. I attempted to mimick the way in which it was done elsewhere within the folding function. Hopefully, that's what you had in mind. Thanks for the suggestion!

andreamari commented on mitiq/zne/scaling/layer_scaling.py:
```diff
@@ -0,0 +1,78 @@
+
+def get_layer_folding(
+    layer_index: int,
+) -> Callable[[QPROGRAM, float], QPROGRAM]:
+    @noise_scaling_converter
```
Docstring missing.

vprusso replied:
Added.

andreamari commented on mitiq/zne/scaling/layer_scaling.py:
```diff
@@ -0,0 +1,78 @@
+    layer_index: int,
+) -> Callable[[QPROGRAM, float], QPROGRAM]:
+    @noise_scaling_converter
+    def fold_ith_layer(
+        circuit: cirq.Circuit, scale_factor: int
+    ) -> cirq.Circuit:
```
Docstring missing.

vprusso replied:
Added.

andreamari commented on mitiq/zne/tests/test_zne.py:
```diff
@@ -492,6 +493,34 @@ def test_execute_with_zne_with_supported_circuits(circuit_type):
+    layer_to_fold = 0
+    fold_layer_func = get_layer_folding(layer_to_fold)
+
+    zne_value = execute_with_zne(
+        circuit, generic_executor, factory=fac, scale_noise=fold_layer_func
+    )
```
Nice to see this is compatible with `execute_with_zne` !

vprusso replied:
Indeed! Thanks to @natestemen !

andreamari commented on docs/source/guide/zne-3-options.md:
````diff
@@ -216,6 +216,72 @@ folding method. Fidelity values must be between zero and one.
+```{code-cell} ipython3
+print(layer_folding(circuit, [2, 3]))
+```
+
````
We should mention that `layer_folding` is not a standard noise scaling function which can be directly used with  `execute_with_zne`  since its signature is not compatible.
 
We could add that:

 `layer_folding` could be used for noise diagnostics `cite Calderon_2023_Quantum` or for a low-level implementation of ZNE in a similar way as shown in the `[next section](zne-4-low-level.md)`.
To get a noise scaling function which is instead directly compatible with execute_with_zne one can use  `{func}.get_layer_folding`.

vprusso replied:
I like it. Added!

andreamari commented on mitiq/zne/scaling/layer_scaling.py:
```diff
@@ -0,0 +1,78 @@
+    def fold_ith_layer(
+        circuit: cirq.Circuit, scale_factor: int
+    ) -> cirq.Circuit:
+        layers = [0] * len(circuit)
+        layers[layer_index] = scale_factor
```
This change is important. It maps the scale_factor to the right number of foldings. 

```suggestion
        layers = [0] * len(circuit)
        num_folds = (scale_factor - 1)  /  2
        if np.isclose(num_folds, int(num_folds)):
            num_folds = int(num_folds)
        layers[layer_index] = num_folds
```

The following part allows us using `scale_factor` equal to 1.0, 3.0 , 5.0` instead of just 1, 3, 5):
```
        if np.isclose(num_folds, int(num_folds)):
            num_folds = int(num_folds)
```
It can be removed if you don't like it.

vprusso replied:
I like it! Applied.

andreamari reviewed (commented):
I like the addition of the new function and the removal of `fold_all_layers`. Additional  comments below.

vprusso commented:
Thanks again for the round of comments, @andreamari ! Hopefully, the updates there make sense, and of course, let me know if not. 

@natestemen if you want to give the PR another look over as we briefly chatted about, have at it!

andreamari commented on mitiq/zne/scaling/layer_scaling.py:
```diff
@@ -0,0 +1,115 @@
+    """Applies a variable amount of folding to select layers of a circuit.
+
+    Args:
+        circuit: The input cirq circuit.
```
```suggestion
        circuit: The input circuit.
```
The function is now multi-platform

andreamari commented on mitiq/zne/scaling/layer_scaling.py:
```diff
@@ -0,0 +1,115 @@
+    Args:
+        circuit: The input cirq circuit.
+        layers_to_fold: A list with the index referring to the layer number,
+                        and the element filled by an integer represents the
```
```suggestion
                        and the element filled by an integer representing the
```

andreamari commented on mitiq/zne/scaling/layer_scaling.py:
```diff
@@ -0,0 +1,115 @@
+                        number of times the layer is folded.
+
+    Returns:
+        A cirq ``Circuit`` with layers and number of times to fold specified
+        by ``layers_to_invert``.
```
```suggestion
        The folded circuit.
```

andreamari commented on mitiq/zne/scaling/layer_scaling.py:
```diff
@@ -0,0 +1,115 @@
+
+        _check_foldable(circuit)
+
+        folded = deepcopy(circuit)
+        measurements = _pop_measurements(folded)
+
+        folded = layer_folding(folded, layers_to_fold=layers)
+
+        _append_measurements(folded, measurements)
```
```suggestion
        folded = layer_folding(folded, layers_to_fold=layers)
```
Isn't better to move the following lines:
```
        folded = deepcopy(circuit)
        measurements = _pop_measurements(folded)
        ...body of the function
        _append_measurements(folded, measurements)
```
into the low-level function `layer_folding()` instead of the high-level function `fold_ith_layer`? In this way they would be applied in both cases.

andreamari commented on docs/source/guide/zne-3-options.md:
```diff
@@ -216,6 +216,79 @@ folding method. Fidelity values must be between zero and one.
+as shown in the `[next section](zne-4-low-level.md)`.
+
+To get a noise scaling function which is instead directly compatible with
+execute_with_zne one can use  `{func}.get_layer_folding`.
```
```suggestion
`{func}.execute_with_zne` one can use  `{func}.get_layer_folding`.
```

andreamari commented on mitiq/zne/scaling/layer_scaling.py:
```diff
@@ -0,0 +1,115 @@
+
+    if len(layers_to_fold) != len(circuit):
+        raise ValueError(
+            "Length of `layers_to_fold` must be equal to length of circuit."
```
```suggestion
            f"Length of layers_to_fold must be equal to the circuit depth excluding measurements, i.e., {len(circuit_without_meas)})"
```
Up to style, can we say something like this? Note that this suggestion assumes `circuit_without_meas` has been created.

A reason to provide the exact depth is that the user may not know what is the length of the internal circuit representation, since a conversion from say Qiskit to Cirq could change the depth.

andreamari commented on mitiq/zne/scaling/layer_scaling.py:
```diff
@@ -0,0 +1,115 @@
+        A cirq ``Circuit`` with layers and number of times to fold specified
+        by ``layers_to_invert``.
+    """
+    circuit = cirq.synchronize_terminal_measurements(circuit)
```
```suggestion
    circuit = cirq.synchronize_terminal_measurements(circuit)
```
This seems dangerous since it could mutate the input variable `circuit`. But, if it doesn't, ignore this comment.

vprusso replied:
Right. I think this did make sense before but perhaps now can be taken out. Removed!

andreamari reviewed (commented):
A few other comments. I promise that, in the next round, I'll approve!  :-D

vprusso commented:
> A few other comments. I promise that, in the next round, I'll approve! :-D

Thank you for the comments, @andreamari , and thank you for the suggested fixes. Of course, don't approve if you think it's still lacking in some way, but hopefully this is closer to ready in any case!

vprusso commented:
After a bit of wrestling with the tests, I think we are good!
https://github.com/unitaryfund/mitiq/actions/runs/4760447950

andreamari commented:
> After a bit of wrestling with the tests, I think we are good! https://github.com/unitaryfund/mitiq/actions/runs/4760447950

andreamari commented:
I think you missed the text of this [comment](https://github.com/unitaryfund/mitiq/pull/1767#discussion_r1172408568) in which the suggestion was to move some lines of code from the `fold_ith_layer` function into the `layer_folding` function. They instead completely disappeared. 

Moreover, it would be good to have back the error message that was removed in the recent commit [bfa9389](https://github.com/unitaryfund/mitiq/pull/1767/commits/bfa9389782994267cb899964c79dccd971a68e99). That message is useful to the user in order to understand how long `layers_to_fold` should be.

If I am missing some aspect or if there are any problems, we can sync on Discord.

natestemen commented on docs/source/refs.bib:
```diff
@@ -88,6 +88,16 @@ @article{Cafaro_2014_PRA
   url       = {https://link.aps.org/doi/10.1103/PhysRevA.89.022316},
 }
 
+@article{Calderon_2023_Quantum,
+  title       = {Quantum circuit debugging and sensitivity analysis via local inversions},
+  author      = {Calderon-Vargas, Fernando A and Proctor, Timothy and Rudinger, Kenneth and Sarovar, Mohan},
+  journal     = {Quantum},
+  volume      = {7},
+  pages       = {921},
+  year        = {2023},
+  publisher   = {Verein zur F{\"o}rderung des Open Access Publizierens in den Quantenwissenschaften},
+}
```
Adding a URL (perhaps https://doi.org/10.22331/q-2023-02-09-921) would be helpful here to make the bibliography more useful.

vprusso replied:
Done!

natestemen commented on docs/source/guide/zne-3-options.md:
````diff
@@ -216,6 +216,79 @@ folding method. Fidelity values must be between zero and one.
+print(layer_folding(circuit, [2, 3]))
+```
+
+The `layer_folding` function could be used for noise diagnostics `cite
+Calderon_2023_Quantum` or for a low-level implementation of ZNE in a similar way
+as shown in the `[next section](zne-4-low-level.md)`.
````
```suggestion
The `layer_folding` function could be used for noise diagnostics {cite}`Calderon_2023_Quantum`
or for a low-level implementation of ZNE in a similar way
as shown in the [next section](zne-4-low-level.md).
```

I'll let you determine how to break lines, but the citation and "next section" links are broken.

natestemen commented on docs/source/guide/zne-3-options.md:
```diff
@@ -216,6 +216,79 @@ folding method. Fidelity values must be between zero and one.
+Calderon_2023_Quantum` or for a low-level implementation of ZNE in a similar way
+as shown in the `[next section](zne-4-low-level.md)`.
+
+To get a noise scaling function which is instead directly compatible with
+`{func}.execute_with_zne` one can use  `{func}.get_layer_folding`.
```
```suggestion
To get a noise scaling function which is instead directly compatible with
{func}`.execute_with_zne` one can use  {func}`.get_layer_folding`.
```

**untested**, but I think this is the syntax.

natestemen commented on mitiq/zne/scaling/layer_scaling.py:
```diff
@@ -0,0 +1,107 @@
+def get_layer_folding(
+    layer_index: int,
+) -> Callable[[QPROGRAM, float], QPROGRAM]:
+    """Applies a variable amount of folding to select layers of a circuit.
+
+    Args:
+        layer_index: The layer of the circuit to apply folding to.
+
+    Returns:
+        fold_ith_layer: The function for folding the ith layer.
+    """
```
This function does not apply the folding, rather it returns the function to do folding. I think the docstring could be clarified, and might be worth mentioning this is intended to be used in something like ZNE. Doesn't need to be long.

vprusso replied:
I put in a concise description. Let me know if that's not what you had in mind.

natestemen commented on mitiq/zne/scaling/layer_scaling.py:
```diff
@@ -0,0 +1,107 @@
+            layer_folding: the folded quantum circuit.
+        """
+        layers = [0] * len(circuit)
+        num_folds = (scale_factor - 1) / 2
+        if np.isclose(num_folds, int(num_folds)):
+            num_folds = int(num_folds)
+        layers[layer_index] = int(num_folds)
```
It looks like `num_folds` is getting cast as an int regardless of the situation here. Is that intended? If so, lines 97/98 can likely be cut.

andreamari replied:
Good catch! 
I think line 99 should be `layers[layer_index] = num_folds`  without the `int`. 
The idea is that inputs like `num_folds = 3` or  ` num_folds = 3.0 ` will be cast to `int` , and so the function will work.
On the other hand, something like `num_folds = 3.1` won't be cast to an int and so an error will occur as expected.

vprusso replied:
Makes sense. Changed!

natestemen commented on mitiq/zne/scaling/layer_scaling.py:
```diff
@@ -0,0 +1,107 @@
+            num_folds = int(num_folds)
+        layers[layer_index] = int(num_folds)
+
+        _check_foldable(circuit)
```
Might as well do this before everything else. I wouldn't expect it to affect timing since the above computation should be fast, but conceptually we should check foldability first.

vprusso replied:
Definitely makes sense. Changed!

natestemen commented on mitiq/zne/tests/test_zne.py:
```diff
@@ -31,6 +31,7 @@
     fold_gates_at_random,
     insert_id_layers,
 )
+from mitiq.zne.scaling.layer_scaling import get_layer_folding
```
If we add `get_layer_folding` to the `scaling/__init__.py` file, we can add this import to the group directly above.

vprusso replied:
Good call. Changed!

natestemen commented on mitiq/zne/tests/test_zne.py:
```diff
@@ -492,6 +493,34 @@ def test_execute_with_zne_with_supported_circuits(circuit_type):
+        circuit, generic_executor, factory=fac, scale_noise=fold_layer_func
+    )
+
+    # Test zero noise limit is better than unmitigated expectation value
+    assert abs(unmitigated - expected) > abs(zne_value - expected)
```
I was a little worried this might not be deterministic enough, but I ran the test 2k times and it didn't fail, so that's good enough for me!

vprusso replied:
Haha, appreciate the check!

natestemen commented on mitiq/zne/scaling/tests/test_layer_scaling.py:
```diff
@@ -0,0 +1,88 @@
+    LineQubit,
+    ops,
+)
+from mitiq.zne.scaling.layer_scaling import layer_folding
```
```suggestion
from mitiq.zne.scaling import layer_folding
```

This worked for me!

natestemen reviewed (approved):
Awesome work Vincent. Very complete addition to Mitiq. I hope adding the `get_layer_folding` function makes your research code a little cleaner.

I left a few comments, but happy for this to go out once addressed without an additional round of feedback.

vprusso commented:
Final ping to @natestemen before merging. I believe all of the above items have been addressed but wanted to ensure before pressing the big green button.

natestemen commented on mitiq/zne/scaling/layer_scaling.py:
```diff
@@ -0,0 +1,109 @@
+    Args:
+        layer_index: The layer of the circuit to apply folding to.
+
+    Returns:
+        fold_ith_layer: The function for folding the ith layer.
```
```suggestion
    Returns:
        The function for folding the ith layer.
```

natestemen commented on mitiq/zne/scaling/layer_scaling.py:
```diff
@@ -0,0 +1,109 @@
+            circuit: Circuit to fold.
+            scale_factor: Factor to scale the circuit by.
+
+        Returns:
+            layer_folding: the folded quantum circuit.
```
```suggestion
        Returns:
            The folded quantum circuit.
```

natestemen reviewed (approved):
Congrats again Vincent! 🎉

Left two minor comments to help fix the API-doc. Also, no need to tell the user the internal variable name we use for the returned object.

| before | after |
| ------ | ----- |
| <img width="576" alt="Screenshot 2023-04-28 at 2 43 34 PM" src="https://user-images.githubusercontent.com/12703123/235259521-578d7cde-e29e-4ddc-859b-9f9cf74bcf10.png"> | <img width="797" alt="Screenshot 2023-04-28 at 2 43 40 PM" src="https://user-images.githubusercontent.com/12703123/235259551-3da0ce5c-785d-486c-9a29-5c4c6b3afdc6.png"> |

vprusso commented:
Awesome! Merging post-changes. Thanks again for all of the help @natestemen and @andreamari !

vprusso commented:
Build is failing but it seems to be something unrelated to the contributions of this diff? Any thoughts @andreamari @natestemen. Want to avoid pushing broken code

natestemen commented:
Looks like an unrelated failure, but I've triggered it to build again to ensure this is the case.

andreamari commented:
Tests are green now and so we can finally merge! 
Congrats @vprusso!  :partying_face:

vprusso commented:
Woo-hoo! Thank you for all the help again, @andreamari and @natestemen !
